### PR TITLE
Initial implementation for DBConnection.GetSchema(String) #443

### PR DESCRIPTION
**Pull request by Thorium**

### CODE REVIEW

**Comment by Thorium** [comment]
The Appveyor test fails as I was naive and used the test database file `readonly.db` that was used by the other tests. However I tested locally with larger database and it works. How should the databases be used in the unit tests?

**Comment by bricelam** [comment]
> How should the databases be used in the unit tests?

Use `:memory:` unless the test requires otherwise.

File: test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs [hunk 1]
```diff
@@ -890,5 +890,26 @@ public void DataChange_event_works()
                 Assert.Equal(101, list[0].RowId);
             }
         }
+
+        [Fact]
+        public void GetSchema_has_collections()
+        {
+            var connectionString = "Data Source=readonly.db";
```
Comment (smitpatel) [inline test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs]: try datasource as `:memory:` ?

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs [hunk 2]
```diff
@@ -556,5 +556,138 @@ public AggregateContext(T seed)
...
+        /// Returns schema information for the data source of this DbConnection using the specified string for the schema name.
+        /// </summary>
+        /// <param name="collectionName">Specifies the name of the schema to return.</param>
+        /// <returns>A DataTable that contains schema information.</returns>
+        public override System.Data.DataTable GetSchema(string collectionName)
+        {
+            var dt = new DataTable(collectionName);
+            switch (collectionName)
+            {
+                case "DataTypes":
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: Use `DbMetaDataCollectionNames.DataTypes`

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs [hunk 3]
```diff
@@ -556,5 +556,138 @@ public AggregateContext(T seed)
...
+                    dt.Columns.AddRange(new[] {
+                        new DataColumn("DataType", typeof(string)),
+                        new DataColumn("TypeName", typeof(string)),
+                        new DataColumn("ProviderDbType", typeof(int))
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: Use `DbMetaDataColumnNames.DataType`, etc.

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs [hunk 4]
```diff
@@ -556,5 +556,138 @@ public AggregateContext(T seed)
...
+                        new DataColumn("TABLE_NAME"),
+                        new DataColumn("FKEY_TO_CATALOG"),
+                        new DataColumn("TABLE_CATALOG"),
+                        new DataColumn("FKEY_TO_TABLE"),
+                        new DataColumn("FKEY_FROM_COLUMN"),
+                        new DataColumn("FKEY_TO_COLUMN"),
+                        new DataColumn("CONSTRAINT_NAME")
+                    });
+                    foreach(var tablename in tables){
+                        var relationQuery = "pragma foreign_key_list(" + tablename + ")";
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: Can combine this and the previous query using this syntax:
``` sql
SELECT t.name,
       fk.table,
       fk.from,
       fk.to
FROM sqlite_master t,
     pragma_foreign_key_list(t.name) fk
WHERE t.type = 'table'
```

**Comment by bricelam** [comment]
We should probably implement the *MetaDataCollections* collection to show which collections we currently support.

For the collections we do support, we should probably include all the "usual" columns (even if we don't populate them with data).

**Comment by bricelam** [comment]
Let's add a few more tests to make sure the data returned is correct.

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs (hunk 4 above)
Comment (Thorium) [reply src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: Nice idea, however [DB Browser for SQLite](http://sqlitebrowser.org/) doesn't support that, so I don't know how widely that is supported.

**Comment by Thorium** [comment]
@bricelam MetaDataCollections and more testing added.

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs [hunk 5]
```diff
@@ -556,5 +556,162 @@ public AggregateContext(T seed)
...
+
+        /// <summary>
+        /// System.Data.Common, initial implementation of API DBConnection.GetSchema(String)
+        /// Returns schema information for the data source of this DbConnection using the specified string for the schema name.
+        /// </summary>
+        /// <param name="collectionName">Specifies the name of the schema to return.</param>
+        /// <returns>A DataTable that contains schema information.</returns>
+        public override System.Data.DataTable GetSchema(string collectionName)
+        {
+            return GetSchema(collectionName, null);
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: I suppose we should also add:
``` csharp
public override DataTable GetSchema()
    => GetSchema(DbMetaDataCollectionNames.MetaDataCollections, null);
```

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs [hunk 6]
```diff
@@ -556,5 +556,162 @@ public AggregateContext(T seed)
...
+                    dt.Rows.Add(new object[] { "System.String","string",16 });
+                    dt.Rows.Add(new object[] { "System.String","char",16 });
+                    dt.Rows.Add(new object[] { "System.String","nchar",16 });
+                    dt.Rows.Add(new object[] { "System.DateTime","datetime",6 });
+                    dt.Rows.Add(new object[] { "System.DateTime","smalldate",6 });
+                    dt.Rows.Add(new object[] { "System.DateTime","timestamp",6 });
+                    dt.Rows.Add(new object[] { "System.DateTime","date",6 });
+                    dt.Rows.Add(new object[] { "System.DateTime","time",6 });
+                    dt.Rows.Add(new object[] { "System.Guid","uniqueidentifier",4 });
+                    dt.Rows.Add(new object[] { "System.Guid","guid",4 });
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: We need to refine this information before merging.

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs (hunk 5 above)
Comment (Thorium) [reply src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: done

File: src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs (hunk 6 above)
Comment (Thorium) [reply src/Microsoft.Data.Sqlite.Core/SqliteConnection.cs]: The information here describes what types of .NET can be mapped to types that will be understood by the database driver, and the corresponding System.Data.Common DbTypes. This information is taken from the System.Data.SQLite.dll library.

I think this is more design decision. Do you want to continue from here?
I'm already happy with the current implementation but feel free to change what ever you need.

**Comment by bricelam** [comment]
Closing this per [our discussion](https://github.com/aspnet/Microsoft.Data.Sqlite/issues/443#issuecomment-339444178). Querying directly for this metadata (and hard-coding a list of supported types) is the correct way to handle this for the time being.

**Pull request closed without merging**